jgclark.NoteHelpers/src/unlinkedNoteFinder.js

```diff
@@ -0,0 +1,186 @@
+
+/**
+ * Find all unlinked notes in the current note and create [[links]] to them
+ */
+export async function findUnlinkedNotesInCurrentNote() {
+  const currentNote = Editor.note
+  if (currentNote) {
+    await findUnlinkedNotes([currentNote])
+  }
+}
+
+/**
+ * Find all unlinked notes in all notes and create [[links]] to them
+ */
+export async function findUnlinkedNotesInAllNotes() {
+  const runTime = new Date()
+  const allNotes = [...DataStore.projectNotes, ...DataStore.calendarNotes]
+  const foundLinks = await findUnlinkedNotes(allNotes)
+  logInfo(`Found ${foundLinks} unlinked notes in all notes, took: ${(new Date().getTime() - runTime.getTime()) / 60}s`)
+}
+
+/**
+ * Find all unlinked notes in the given notes
+ * @param {Array<TNote>} notes - the notes to search for unlinked notes in
+ * @returns {number} - the number of unlinked notes found
+ **/
+async function findUnlinkedNotes(notes: Array<TNote>): Promise<number> {
+  let foundLinks = 0
+  try {
+    await CommandBar.onAsyncThread()
+    CommandBar.showLoading(true, 'Finding unlinked notes')
+
+    foundLinks = notes.reduce((count, note) => count + findUnlinkedNotesInNote(note), 0)
+
+    CommandBar.showLoading(false)
+    CommandBar.onMainThread()
+  } catch (error) {
+    logError(pluginJson, JSP(error))
+  }
+  return foundLinks
+}
+
+/**
+ * Find all note titles in the current note and replace them with [[note title]]
+ * @param {TNote} currentNote - the note to search for links in
+ * @returns {number} - the number of links found
+ */
+function findUnlinkedNotesInNote(currentNote: TNote): number {
+  let foundLinks = 0
+  const overallTime = new Date()
+  let content = currentNote.content ?? ''
+
+  logDebug(`Searching for unlinked notes in: ${currentNote.title ?? ''}`)
+
+  const [contentWithCodeBlocksRemoved, codeBlockTracker] = extractCodeBlocks(content)
+  content = contentWithCodeBlocksRemoved
+
+  const [contentWithLinksRemoved, markdownLinkTracker] = extractMarkdownLinks(content)
+  content = contentWithLinksRemoved
+
+  getAllNoteTitlesSortedByLength().forEach((note) => {
+    if (currentNote.title !== note && content.includes(note)) {
+      content = content.replaceAll(buildRegex(note), (_) => {
+        logDebug(`Found link to: ${note}`)
+        foundLinks++
+        return `[[${note}]]`
+      })
+    }
+  })
+
+  content = replaceCodeBlocks(content, codeBlockTracker)
+  content = replaceMarkdownLinks(content, markdownLinkTracker)
+
+  if (foundLinks > 0) {
+    currentNote.content = content
+  }
+  logInfo(`Linked ${foundLinks} notes in ${currentNote.title ?? ''}, took: ${new Date().getTime() - overallTime.getTime()}ms`)
+
+  return foundLinks
+}
+
+/**
+ * A regex pattern to search for note titles in a note.
+ *
+ * \\w*: Matches any word character (alphanumeric + underscore) zero or more times.
+ * (?<!\\[{2}[^[\\]]*): A negative lookbehind that asserts what precedes is not two open square brackets followed by any characters except closing square brackets.
```

Comment:
I love complex regex like this ... but they can also be a nightmare when you find a new edge case.

My memory says that negative lookbehind (or perhaps negative lookahead) doesn't work in Apple's JS regex engine. I've certainly had to rewrite complex regex in the past to avoid one or the other. 
Things might have changed, of course, but it's just a note to make sure you're testing the negative ones thoroughly.

There are also some advanced character classes available in more recent regex (including the Apple JS one) that allow full unicode to work.  e.g. in Search Extensions I've used full unicode letter characters (\p{L}) and numbers (\p{N}) rather than ASCII (\w).

Reply:
Ive optimized this see coming pr.